File: app/src/main/java/com/dpis/module/settings/AppLocaleManager.java
package com.dpis.module.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.LocaleList;

import com.dpis.module.R;

import java.util.List;

public final class AppLocaleManager {
    public static final String TAG_FOLLOW_SYSTEM = "";
    public static final String TAG_ENGLISH = "en";
    public static final String TAG_SIMPLIFIED_CHINESE = "zh-CN";
    public static final String TAG_JAPANESE = "ja-JP";
    public static final String TAG_RUSSIAN = "ru";

    private static final String PREFS_NAME = "app_locale";
    private static final String KEY_LANGUAGE_TAG = "language_tag";
    private static final List<LanguageOption> SUPPORTED_LANGUAGES = List.of(
            new LanguageOption(TAG_FOLLOW_SYSTEM, R.string.settings_language_follow_system),
            new LanguageOption(TAG_ENGLISH, R.string.settings_language_english),
            new LanguageOption(TAG_SIMPLIFIED_CHINESE, R.string.settings_language_simplified_chinese),
            new LanguageOption(TAG_JAPANESE, R.string.settings_language_japanese),
            new LanguageOption(TAG_RUSSIAN, R.string.settings_language_russian));

    public static final class LanguageOption {
        public final String tag;
        public final int labelResId;

        private LanguageOption(String tag, int labelResId) {
            this.tag = tag;
            this.labelResId = labelResId;
        }
    }

    private AppLocaleManager() {
    }

    public static Context wrap(Context context) {
        String languageTag = getLanguageTag(context);
        if (languageTag.isEmpty()) {
            return context;
        }
        java.util.Locale locale = java.util.Locale.forLanguageTag(languageTag);
        Configuration configuration = new Configuration(context.getResources().getConfiguration());
        configuration.setLocales(new LocaleList(locale));
        configuration.setLocale(locale);
        return context.createConfigurationContext(configuration);
    }

    public static String getLanguageTag(Context context) {
        return getPreferences(context).getString(KEY_LANGUAGE_TAG, TAG_FOLLOW_SYSTEM);
    }

    public static boolean setLanguageTag(Context context, String languageTag) {
        return getPreferences(context)
                .edit()
                .putString(KEY_LANGUAGE_TAG, sanitizeLanguageTag(languageTag))
                .commit();
    }

    public static List<LanguageOption> supportedLanguages() {
        return SUPPORTED_LANGUAGES;
    }

    public static int selectedLabelResId(Context context) {
        return selectedLanguage(context).labelResId;
    }

    public static LanguageOption selectedLanguage(Context context) {
        String languageTag = getLanguageTag(context);
        for (LanguageOption option : SUPPORTED_LANGUAGES) {
            if (option.tag.equals(languageTag)) {
                return option;
            }
        }
        return SUPPORTED_LANGUAGES.get(0);
    }

    private static String sanitizeLanguageTag(String languageTag) {
        for (LanguageOption option : SUPPORTED_LANGUAGES) {
            if (option.tag.equals(languageTag)) {
                return option.tag;
            }
        }
        return TAG_FOLLOW_SYSTEM;
    }

    private static SharedPreferences getPreferences(Context context) {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }
}

File: app/src/main/java/com/dpis/module/settings/AppLocaleManager.kt
package com.dpis.module.settings

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.os.LocaleList
import com.dpis.module.R

/** Owns the explicit app language choice and creates configuration-localized contexts. */
object AppLocaleManager {
    @JvmField val TAG_FOLLOW_SYSTEM = ""
    @JvmField val TAG_ENGLISH = "en"
    @JvmField val TAG_SIMPLIFIED_CHINESE = "zh-CN"
    @JvmField val TAG_JAPANESE = "ja-JP"
    @JvmField val TAG_RUSSIAN = "ru"

    private const val PREFS_NAME = "app_locale"
    private const val KEY_LANGUAGE_TAG = "language_tag"
    private val supportedLanguages = listOf(
        LanguageOption(TAG_FOLLOW_SYSTEM, R.string.settings_language_follow_system),
        LanguageOption(TAG_ENGLISH, R.string.settings_language_english),
        LanguageOption(TAG_SIMPLIFIED_CHINESE, R.string.settings_language_simplified_chinese),
        LanguageOption(TAG_JAPANESE, R.string.settings_language_japanese),
        LanguageOption(TAG_RUSSIAN, R.string.settings_language_russian),
    )

    class LanguageOption(
        @JvmField val tag: String,
        @JvmField val labelResId: Int,
    )

    @JvmStatic
    fun wrap(context: Context): Context {
        val languageTag = getLanguageTag(context)
        val configuration = Configuration(context.resources.configuration)
        if (languageTag.isEmpty()) {
            // A recreated Activity can retain its previous app language. Follow-system must
            // explicitly restore the device locale instead of inheriting that stale context.
            val systemLocales = Resources.getSystem().configuration.locales
            configuration.setLocales(systemLocales)
            if (!systemLocales.isEmpty()) {
                configuration.setLocale(systemLocales[0])
            }
        } else {
            val locale = java.util.Locale.forLanguageTag(languageTag)
            configuration.setLocales(LocaleList(locale))
            configuration.setLocale(locale)
        }
        return context.createConfigurationContext(configuration)
    }

    @JvmStatic
    fun getLanguageTag(context: Context): String =
        preferences(context).getString(KEY_LANGUAGE_TAG, TAG_FOLLOW_SYSTEM) ?: TAG_FOLLOW_SYSTEM

    @JvmStatic
    fun setLanguageTag(context: Context, languageTag: String): Boolean =
        preferences(context).edit().putString(KEY_LANGUAGE_TAG, sanitizeLanguageTag(languageTag)).commit()

    @JvmStatic
    fun supportedLanguages(): List<LanguageOption> = supportedLanguages

    @JvmStatic
    fun selectedLabelResId(context: Context): Int = selectedLanguage(context).labelResId

    @JvmStatic
    fun selectedLanguage(context: Context): LanguageOption =
        supportedLanguages.firstOrNull { it.tag == getLanguageTag(context) }
            ?: supportedLanguages.first()

    private fun sanitizeLanguageTag(languageTag: String?): String =
        supportedLanguages.firstOrNull { it.tag == languageTag }?.tag ?: TAG_FOLLOW_SYSTEM

    private fun preferences(context: Context) =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
